src/lib/core/assoc.h

```diff
@@ -189,6 +189,76 @@ mh_strnu32_find_str(struct mh_strnu32_t *h, const char *str, uint32_t len)
 	return mh_strnu32_find(h, &key, NULL);
 }
 
+/*
+ * Map: (const char *, uint32_t, const char *, uint32_t) => (void *)
+ */
+
+static inline uint32_t
+mh_access_hash(const char *uname, uint32_t uname_len,
+	       const char *fname, uint32_t fname_len)
+{
+	return mh_strn_hash(uname, uname_len) ^ mh_strn_hash(fname, fname_len);
+}
+
+#define mh_name _access
+
+/**
+ * Key of `mh_access_node_t` hash table.
+ */
+struct mh_access_key_t {
+	/* Key user name. */
+	const char *uname;
+	/* Key user name length. */
+	uint32_t uname_len;
+	/* Key function name. */
+	const char *fname;
+	/* Key function name length. */
+	uint32_t fname_len;
+	/* Key hash calculated using `mh_access_hash`. */
+	uint32_t hash;
+};
+
+#define mh_key_t struct mh_access_key_t *
+
+/**
+ * Node of `mh_access_node_t` hash table.
+ */
+struct mh_access_node_t {
+	/* Key user name. */
+	const char *uname;
+	/* Key user name length. */
+	uint32_t uname_len;
+	/* Key function name. */
+	const char *fname;
+	/* Key function name length. */
+	uint32_t fname_len;
+	/* Key hash calculated using `mh_access_hash`. */
+	uint32_t hash;
+	/* Mapped value. */
+	int val;
+};
+
+#define mh_node_t struct mh_access_node_t
```

Comment:
It seems, we don't need a separate key/node structs. It is enough for us to have a set, without a mapped value.

At least, the following change seems to work.

```diff
diff --git a/src/box/rt_lua_call_access.c b/src/box/rt_lua_call_access.c
index d17fc63eb6..0fc486bf83 100644
--- a/src/box/rt_lua_call_access.c
+++ b/src/box/rt_lua_call_access.c
@@ -50,11 +50,9 @@ grant_rt_access(const char *uname, uint32_t uname_len,
 	/* Grant access to lua_call function. */
 	uint32_t hash = mh_access_hash(uname, uname_len, fname, fname_len);
 	const struct mh_access_node_t access_node =
-		{ uname, uname_len, fname, fname_len, hash, RT_ACCESS };
-	struct mh_access_node_t repl_access_node = { NULL, 0, NULL, 0, 0, 0 };
-	struct mh_access_node_t *prepl_access_node = &repl_access_node;
-	mh_access_put(user_rt_access_cache, &access_node, &prepl_access_node,
-		      NULL);
+		{ uname, uname_len, fname, fname_len, hash };
+	struct mh_access_node_t *ignored;
+	mh_access_put(user_rt_access_cache, &access_node, &ignored, NULL);
 }
 
 void
@@ -72,7 +70,7 @@ revoke_rt_access(const char *uname, uint32_t uname_len,
 		/* Revoke access to lua_call function. */
 		uint32_t hash = mh_access_hash(uname, uname_len, fname,
 					       fname_len);
-		struct mh_access_key_t key =
+		struct mh_access_node_t key =
 			{ uname, uname_len, fname, fname_len, hash };
 		mh_int_t pos = mh_access_find(user_rt_access_cache, &key, NULL);
 		if (pos != mh_end(user_rt_access_cache))
@@ -92,7 +90,7 @@ check_rt_access(const char *uname, uint32_t uname_len,
 		return RT_ACCESS;
 
 	uint32_t hash = mh_access_hash(uname, uname_len, fname, fname_len);
-	struct mh_access_key_t key =
+	struct mh_access_node_t key =
 		{ uname, uname_len, fname, fname_len, hash };
 	pos = mh_access_find(user_rt_access_cache, &key, NULL);
 	if (pos == mh_end(user_rt_access_cache))
diff --git a/src/lib/core/assoc.h b/src/lib/core/assoc.h
index 2b2b0ec310..5f63a47776 100644
--- a/src/lib/core/assoc.h
+++ b/src/lib/core/assoc.h
@@ -190,7 +190,7 @@ mh_strnu32_find_str(struct mh_strnu32_t *h, const char *str, uint32_t len)
 }
 
 /*
- * Map: (const char *, uint32_t, const char *, uint32_t) => (void *)
+ * Set: (const char *, uint32_t, const char *, uint32_t)
  */
 
 static inline uint32_t
@@ -202,24 +202,6 @@ mh_access_hash(const char *uname, uint32_t uname_len,
 
 #define mh_name _access
 
-/**
- * Key of `mh_access_node_t` hash table.
- */
-struct mh_access_key_t {
-	/* Key user name. */
-	const char *uname;
-	/* Key user name length. */
-	uint32_t uname_len;
-	/* Key function name. */
-	const char *fname;
-	/* Key function name length. */
-	uint32_t fname_len;
-	/* Key hash calculated using `mh_access_hash`. */
-	uint32_t hash;
-};
-
-#define mh_key_t struct mh_access_key_t *
-
 /**
  * Node of `mh_access_node_t` hash table.
  */
@@ -234,10 +216,9 @@ struct mh_access_node_t {
 	uint32_t fname_len;
 	/* Key hash calculated using `mh_access_hash`. */
 	uint32_t hash;
-	/* Mapped value. */
-	int val;
 };
 
+#define mh_key_t struct mh_access_node_t *
 #define mh_node_t struct mh_access_node_t
 
 #define mh_arg_t void *
@@ -255,7 +236,7 @@ mh_access_find_str(struct mh_access_t *h, const char *uname, uint32_t uname_len,
 		   const char *fname, uint32_t fname_len)
 {
 	uint32_t hash = mh_access_hash(uname, uname_len, fname, fname_len);
-	struct mh_access_key_t key = {uname, uname_len, fname, fname_len, hash};
+	struct mh_access_node_t key = {uname, uname_len, fname, fname_len, hash};
 	return mh_access_find(h, &key, NULL);
 }
 
```

Reply:
Done.